default/python/universe_generation/natives.py

```diff
@@ -76,6 +76,10 @@ def generate_natives(native_freq, systems, empire_home_systems):  # noqa: max-co
                     natives_for_planet_type[planet_type].append(species)
                     planet_types_for_natives[species].add(planet_type)
 
+    # create a list to store native species already picked. This is used to reduce number of duplicates
```

Comment:
We could use the Python built-in feature to deal with that.  It still allows duplications but will make them controllable.

I use a generator, to avoid extra code that accumulates the result. 


```python
import random


def _build_wights(items: list[str], number_of_items_to_select: int):
    return [number_of_items_to_select for _ in items]


def select_recent(items: list[str], number_of_items_to_select, reduce_factor=1):
    """
    Select the number_of_items_to_select items, each time reducing the probability to select the same item.

    reduce_factor is define the probability to choose the unique item.
    1 is maximum duplication are allowed, >= len(items) no duplication are allowed.

    >>> from collections import Counter
    >>> Counter(select_recent(["a", "b", "c"], 3, reduce_factor=3)) == Counter({"a": 1, "b": 1, "c": 1})

    If factor is > 1 and number_of_items_to_select is bigger than number of items, duplication are still possible.

    For example, this code will produce exactly 2 sets of unique items

     >>> from collections import Counter
    >>> Counter(select_recent(["a", "b", "c"], 6, reduce_factor=3)) == Counter({"a": 2, "b": 2, "c": 2})

    """
    weights = _build_wights(items, number_of_items_to_select)

    for _ in range(number_of_items_to_select):
        try:
            choice = random.choices(items, weights=weights, k=1)[0]
        except ValueError:
            # all weights are zero
            weights = _build_wights(items, number_of_items_to_select)
            choice = random.choices(items, weights=weights, k=1)[0]
        yield choice
        weights[items.index(choice)] = max(weights[items.index(choice)] - reduce_factor, 0)


print(list(select_recent(["a", "b", "c"], 3, reduce_factor=3)))
```

Reply:
This one I plan to rewrite so it knows upfront how many natives its placing, so it can work out how many repeats are necessary